productivity-suite/piercing-library/src/piercing-fragment-outlet.ts

```diff
@@ -50,21 +52,22 @@ export class PiercingFragmentOutlet extends HTMLElement {
       );
     }
 
-    this.fragmentId = fragmentId;
-
-    let fragmentHost = this.getFragmentHost();
+    this.fragmentHost = this.getFragmentHost(fragmentId);
 
-    if (fragmentHost) {
-      this.pierceFragmentIntoDOM(fragmentHost);
+    if (this.fragmentHost) {
+      // There is already a fragment host in the DOM that we can pierce into this outlet
+      this.innerHTML == "";
+      this.fragmentHost.pierceInto(this);
     } else {
-      const fragmentStream = await this.fetchFragmentStream();
-      await this.streamFragmentIntoOutlet(fragmentStream);
-      fragmentHost = this.getFragmentHost(true);
+      // We need to fetch and create the fragment host
+      const fragmentStream = await this.fetchFragmentStream(fragmentId);
+      await this.streamFragmentIntoOutlet(fragmentId, fragmentStream);
+      this.fragmentHost = this.getFragmentHost(fragmentId, true);
```

Comment:
The key point here is that `this.fragmentHost` is only set after the `streamIntoOutlet()` method completed.